src/echo-common/src/cached-task.ts

```diff
@@ -3,84 +3,110 @@ import { throttleTime } from 'rxjs/operators'
 import { filterNullish } from 'ts-ratchet'
 import { EchoDirService } from './echo-dir-service'
 
-export type CachedTaskEvent<T> = {
+export type SmartCacheLoadMode = "Default" | "CacheOnly" | "NoCache"
+
+export type SmartCacheLoadConfig = {
+  key: string,
+  source: string,
+  mode?: SmartCacheLoadMode
+}
+
+export type SmartCacheResultEvent<T> = {
+  type: "result",
+  key: string,
+  result: T | null | undefined,
+  timestampMs: number
+}
+
+export type SmartCacheEvent<T> = SmartCacheResultEvent<T>
+
+export type SmartCacheLoadEvent = {
   key: string
-  state?: string
-  status?: string
-  result?: T | null | undefined
-  timestampMs?: number
 }
 
-export class CachedTask<T> {
-  private tasks$ = new Subject<CachedTaskEvent<T>>()
-  public events$ = new Subject<CachedTaskEvent<T>>()
-  public cache$ = new BehaviorSubject<{ [key: string]: CachedTaskEvent<T> }>({})
+export class SmartCache<T> {
+  private taskQueue$ = new Subject<SmartCacheLoadEvent>()
+  private events$ = new Subject<SmartCacheEvent<T>>()
+
+  //What if there was on BehaviorSubject per key that contained the state {loading, error, data}
```

Comment:
I really like that you accept my idea with the different statuses.
If we have the base, then we can expand this object with more information.

`loading, error, data` should be renamed to `isLoading, isError, data` to be more in line with other libraries. Because error could mean an error object or an error boolean? Its not clear enought, because of that loading should be renamed to `isLoading` that its inline with `isError`
`isFetching` is really important, because most of the time I rely on `isFetching`, because if I would only want to show the initial loading screen on the first fetch not on every other. 
Please read this:
```
Frequently Used Query Hook Return Values
The query hook returns an object containing properties such as the latest data for the query request, as well as status booleans for the current request lifecycle state. Below are some of the most frequently used properties. Refer to [useQuery](https://redux-toolkit.js.org/rtk-query/api/created-api/hooks#usequery) for an extensive list of all returned properties.

data - The latest returned result regardless of hook arg, if present.
currentData - The latest returned result for the current hook arg, if present.
error - The error result if present.
isUninitialized - When true, indicates that the query has not started yet.
isLoading - When true, indicates that the query is currently loading for the first time, and has no data yet. This will be true for the first request fired off, but not for subsequent requests.
isFetching - When true, indicates that the query is currently fetching, but might have data from an earlier request. This will be true for both the first request fired off, as well as subsequent requests.
isSuccess - When true, indicates that the query has data from a successful request.
isError - When true, indicates that the query is in an error state.
refetch - A function to force refetch the query
```